### ISSUE TITLE
Load Testing with Multi Concurrent users for Dynamics 365

### ISSUE THREAD

**Issue by ksatishcrm** [microsoft/playwright#18487]
Using Playwright, we are testing our test cases for Dynamics 365 and with single user we were able to run our test cases successfully.
But we are struck with load testing, when using multiple concurrent users for running our test cases.

We followed below documentation for multiple users authentication.
https://playwright.dev/docs/auth#avoiding-multiple-sessions-per-account-at-a-time

Can anyone share insights on how to implement loading testing with multi concurrent users.

**Comment by aslushnikov** [comment]
@ksatishcrm For load testing, you can just log in for each test. What exactly doesn't work?

**Comment by ksatishcrm** [comment]
We want to run test cases with multiple concurrent users in parallel for our load testing.

**Comment by aslushnikov** [comment]
You can setup logged in storage states for multiple different users in the `globalSetup` hook and then use them when running all your tests in parallel.

**Comment by yury-s** [comment]
> But we are struck with load testing, when using multiple concurrent users for running our test cases.
>
> We followed below documentation for multiple users authentication.
https://playwright.dev/docs/auth#avoiding-multiple-sessions-per-account-at-a-time

If you followed that example you should end up running N workers in parallel with a unique user on each of them. Can you elaborate on what is missing in for you in this setup? Do you want each test case to use multiple user accounts ?

**Comment by ksatishcrm** [comment]
For load testing we have bunch of test cases and want to run them in parallel with multiple users.

Eg: 
User1 :abc1@xyz.com
User2: abc2@xyz.com

User1 and User2 should run all the test cases in parallel and should give a consolidated report for each test case, the time taken to execute for each test case.

can you please share, if you have any code snippet for this scenario.

**Comment by yury-s** [comment]
There is no ready code snippet but there are multiple ways to do that depending on your preferences. E.g. you could pass the user name in an environment variable to each playwright run:

```sh
TEST_USER=abc1@xyz.com npx playwright test
TEST_USER=abc2@xyz.com npx playwright test
```

and then in your test read it as:
```js
const user = process.env.TEST_USER;
```

Or you can have a user [option fixture](https://playwright.dev/docs/test-fixtures#fixtures-options) and two projects parametrized by user name:

```ts
// Declare your options to type-check your configuration.
export type MyOptions = {
  user: string;
};

// Extend base test by providing "user"
export const test = base.extend<MyOptions>({
  defaultItem: ['test@xyz.com', { option: true }],
});

```

Then in the playwright.config.ts:

```ts
const config: PlaywrightTestConfig = {
...
  projects: [
    {
      name: 'user1',
      use: {
        user: 'abc1@xyz.com'
      }
    },
    {
      name: 'user2',
      use: {
        user: 'abc2@xyz.com'
      }
    },
  ]
};
```
and in the tests:

```ts
test.beforeEach(({ user }) => {
  // do user-specific stuff
});
```

After that you can run all your tests with desired concurrency (see [this doc](https://playwright.dev/docs/test-parallel) for more details). Would that work for you?

**Comment by ksatishcrm** [comment]
As suggested by you, we were able to authenticate multiple users using fixture. 
Below are the steps we followed to achieve authentication for multi users.

**Step1**: Created EnvironmentConfig.ts for storing the environment details. 
Service account and password are saved in build pipeline variables.

`export const environmentConfig : string = `{
    "EnvironmentURL": {         
            "https://testorg.crm8.dynamics.com": "UAT",            
},
"UAT" : {
            "ServiceAccountPassword": "__ServiceAccountPassword__",
            "ServiceAccountPassword2": "__ServiceAccountPassword2__",
            "ServiceAccountPassword3": "__ServiceAccountPassword3__",
            "ServiceAccountPassword4": "__ServiceAccountPassword4__",
            "ServiceAccountPassword5": "__ServiceAccountPassword5__",
            "UATUrl": "__OrgUrl__",
            "ServiceAccountUser": "__ServiceAccountUser__",
            "ServiceAccountUser2": "__ServiceAccountUser2__",
            "ServiceAccountUser3": "__ServiceAccountUser3__",
            "ServiceAccountUser4": "__ServiceAccountUser4__",
            "ServiceAccountUser5": "__ServiceAccountUser5__",
            "AppInsightsKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "EnvironmentName": "UAT"
}
}`;`

**Step2**: Created Fixtures.ts file as below.

`import { test as base } from '@playwright/test';
import { environmentConfig } from "../Util/EnvironmentConfig";

let envConfig: any = {};
const config = JSON.parse(environmentConfig);
envConfig = config['UAT'];

export type UserList = {
    user: string;
    password: string;
  };
  
  // Extend base test by providing "user"
   const fixture = base.extend<UserList>({
    user: [envConfig.ServiceAccountUser, { option: true }],
    password: [envConfig.ServiceAccountPassword, {option:true}]
  });

export const test=fixture;`

**Step3**: In the playwright.config.ts file imported fixtures and EnvironmentConfig

import { UserList } from './Util/Fixtures';
import { environmentConfig } from "./Util/EnvironmentConfig";

Multi user implementation in the playwright.config.ts

` workers: 1,
    fullyParallel: true,
    projects: [
        {
          name: 'abc1',
          use: { user: envConfig.ServiceAccountUser, password: envConfig.ServiceAccountPassword },          
        },
        {
          name: 'abc2',
          use: { user: envConfig.ServiceAccountUser2, password: envConfig.ServiceAccountPassword2},
        },
        {
          name: 'abc3',
          use: { user: envConfig.ServiceAccountUser3, password: envConfig.ServiceAccountPassword3},
        },
        {
          name: 'abc4',
          use: { user: envConfig.ServiceAccountUser4, password: envConfig.ServiceAccountPassword4},
        },
        {
          name: 'abc5',
          use: { user: envConfig.ServiceAccountUser5, password: envConfig.ServiceAccountPassword5},
        },
      ],`

**Step4**: Created Login.spec.ts to Authenticate multi users.

`import {test} from '../Util/Fixtures';
import {  expect, chromium, Browser, BrowserContext, Selectors,Page } from '@playwright/test';
import { environmentConfig } from "../Util/EnvironmentConfig";

let envConfig: any = {};
const config = JSON.parse(environmentConfig);
envConfig = config['UAT'];

test('Login', async ({ user, password }) => {
 //   console.log(user, password);
       const browser = await chromium.launch();    
    const page = await browser.newPage({recordVideo:{dir:'videos/'}});  
    
    await Promise.all([      
        // This action triggers the request
        page.goto(envConfig.UATOrgUrl)
      ]);
    await page.locator('input[name="login"]').fill(user);
    await page.locator('text=Next').click();
    const userPassword = page.locator('span[role="button"]:has-text("Password")');
    await userPassword.waitFor({ state: 'visible' })
    await userPassword.click();
    await page.locator('id=passwordInput').fill(password);    
    await page.locator('id=submitButton').click();
    await page.locator('text=Yes').click(); 
    await page.context().storageState({ path: 'storageState.json' });    
    await browser.close();   
});`


With single worker processer we are able to run the test cases in sequential order.
But for load testing we want to run test cases in parallel with multiple concurrent users(total=5) with worker processes(total=5).

Can you please let me know how to save the authentication state and pass the token to every single test case to run it in parallel.

Want to know if single machine can handle multiple worker processes or it requires sharding.
For our case if we have to test load testing with 500 concurrent users how many worker processer are required and how to achieve this.

**Comment by yury-s** [comment]
> But for load testing we want to run test cases in parallel with multiple concurrent users(total=5) with worker processes(total=5).

I recommend you read [this document](https://playwright.dev/docs/test-parallel) that explains Playwright Test process model and parallelization. If I understand correctly, you want to
- run the entire test suite for each of the 5 users
- run the tests on 5 workers in parallel
- there is exactly one worker running tests for each user
- the tests run sequentially for each of the users

This would would work with the project per user configuration I suggested above if we had an option to run all projects in parallel and so that there is exactly one worker per project. Unfortunately, there is no such mode at the moment but there is a workaround that you can try:

Place all your load tests into a single test.spec.ts file to guarantee that the tests will be executed by a single worker. Move the login logic into beforeAll as each of the workers will have to repeat the login process for the assigned user. Configure the projects as you did, but remove  `fullyParallel: true` and set the number of workers to the number of user projects or any number bigger than that: `workers: 5`. This way when you run `npx playwright test` playwright will start exactly 5 workers and each of them will run the single test file for each of the projects. This is basically one user per worker. The test file may look something like this:

```ts
// ...
test.use({
  storageState: ({ user }, use) => use(`storageState-${user}.json`)
});

test.beforeAll(async ({ user, password }) => {
 //   console.log(user, password);
    const browser = await chromium.launch();    
    const page = await browser.newPage({
      storageState: undefined, // this is important!
      recordVideo:{dir:'videos/'}
    });  
    
    await Promise.all([      
        // This action triggers the request
        page.goto(envConfig.UATOrgUrl)
      ]);
    await page.locator('input[name="login"]').fill(user);
    await page.locator('text=Next').click();
    const userPassword = page.locator('span[role="button"]:has-text("Password")');
    await userPassword.waitFor({ state: 'visible' })
    await userPassword.click();
    await page.locator('id=passwordInput').fill(password);    
    await page.locator('id=submitButton').click();
    await page.locator('text=Yes').click(); 

    // make sure the file names do not clash between users
    await page.context().storageState({ path: `storageState-${user}.json` }); 

    await browser.close();   
});

test('perf test 1', async ({ page }) => {
  // ...
});

test('perf test 2', async ({ page }) => {
  // ...
});

// ...

```
We are planning to add [projectSetup](https://github.com/microsoft/playwright/issues/12084) in the future and the login code in your case should go there but for one file beforeAll should work just as well.

> Can you please let me know how to save the authentication state and pass the token to every single test case to run it in parallel.

Now that I better understand your goal, I believe you'd be better off putting it in beforeAll as I described. Otherwise you'd have to iterate over all users in the global setup and create `storageState-${user}.json` for each of them. Those files can be specified in the project config which seems redundant:

```ts
projects: [
  {
    name: 'abc1',
    use: {
      user: envConfig.ServiceAccountUser, 
      storageState: `storageState-${envConfig.ServiceAccountUser,}.json`,
      password: envConfig.ServiceAccountPassword 
    },          
  },
  {
    name: 'abc2',
    use: {
      user: envConfig.ServiceAccountUser2,
      storageState: `storageState-${envConfig.ServiceAccountUser2}.json`,
      password: envConfig.ServiceAccountPassword2
    },
  },
  {
    name: 'abc3',
    use: {
      user: envConfig.ServiceAccountUser3,
      storageState: `storageState-${envConfig.ServiceAccountUser3}.json`,
      password: envConfig.ServiceAccountPassword3
    },
  },
  {
    name: 'abc4',
    use: {
      user: envConfig.ServiceAccountUser4,
      storageState: `storageState-${envConfig.ServiceAccountUser4}.json`,
      password: envConfig.ServiceAccountPassword4
    },
  },
  {
    name: 'abc5',
    use: {
      user: envConfig.ServiceAccountUser5,
      storageState: `storageState-${envConfig.ServiceAccountUser5}.json`,
      password: envConfig.ServiceAccountPassword5
    },
  },
],`
```

> Want to know if single machine can handle multiple worker processes or it requires sharding.

Playwright doesn't have any hard limit on the number of workers you can run on a single machine as long as it has sufficient resources. Normally, playwright would start about (# CPU cores / 2) workers unless explicitly configured. But in case of load testing I would expect the major load to be on the server end and you can probably go with a much higher number.

> For our case if we have to test load testing with 500 concurrent users how many worker processer are required and how to achieve this.

You'll need to run an experiment and see how it performs while you are increasing the number of workers per machine. There is no generic solution. You can certainly [shard](https://playwright.dev/docs/api/class-testconfig#test-config-shard) this across multiple machines, just make sure all tests are in one file so that you don't end up having two or more files for the same user running in parallel.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.
